plasmapy/analysis/tests/test_fit_functions.py

```diff
@@ -0,0 +1,517 @@
+
+        params = [2] * len(foo.param_names)
+        params += [5]
+        with pytest.raises(ValueError):
+            foo.param_errors = params
+
+    def test_func(self):
+        """Test the `func` method."""
+        foo = self.ff_class()
+
+        for x in (0, 1.0, np.linspace(10, 30, num=20)):
+            assert np.allclose(
+                foo.func(x, *self._test_params), self.func(x, *self._test_params)
+            )
+
+        x = [4, 5, 6]
+        assert np.allclose(
+            foo.func(x, *self._test_params), self.func(np.array(x), *self._test_params)
+        )
+
+        with pytest.raises(ValueError):
+            foo.func("hello", *self._test_params)
+
+        with pytest.raises(ValueError):
+            params = list(self._test_params)
+            params[0] = "hello"
+            foo.func(5, *params)
+
+    def test_func_err(self):
+        """Test the `func_err` method."""
+        foo = self.ff_class(
+            params=self._test_params, param_errors=self._test_param_errors
+        )
+
+        for x in (0, 1.0, np.linspace(10, 30, num=20)):
+            assert np.allclose(
+                foo.func_err(x),
+                self.func_err(x, self._test_params, self._test_param_errors),
+            )
+
+        x = [4, 5, 6]
+        results = foo.func_err(x, x_err=0.1, rety=True)
+        assert np.allclose(
+            results[0],
+            self.func_err(
+                np.array(x), self._test_params, self._test_param_errors, x_err=0.1
+            ),
+        )
+        assert np.allclose(results[1], self.func(np.array(x), *self._test_params))
+
+        with pytest.raises(ValueError):
+            foo.func_err("hello")
+
+        with pytest.raises(ValueError):
+            foo.func_err(5, x_err="goodbye")
+
+        with pytest.raises(ValueError):
+            foo.func_err(5, x_err=[0.1, 0.1])
+
+    def test_call(self):
+        """Test __call__ behavior."""
+        foo = self.ff_class()
+        foo.params = self._test_params
+        foo.param_errors = self._test_param_errors
+
+        for x in (0, 1.0, np.linspace(10, 30, num=20)):
+            assert np.allclose(foo(x), self.func(x, *self._test_params))
+
+            # also return error
+            y, y_err = foo(x, reterr=True)
+            assert np.allclose(y, self.func(x, *self._test_params))
+            assert np.allclose(
+                y_err, self.func_err(x, self._test_params, self._test_param_errors),
+            )
+
+        x = [4, 5, 6]
+        x_err = 0.05
+        assert np.allclose(foo(x), self.func(np.array(x), *self._test_params))
+        y, y_err = foo(x, x_err=x_err, reterr=True)
+        assert np.allclose(y, self.func(np.array(x), *self._test_params))
+        assert np.allclose(
+            y_err,
+            self.func_err(
+                np.array(x), self._test_params, self._test_param_errors, x_err=x_err
+            ),
+        )
+
+        with pytest.raises(ValueError):
+            foo("hello")
```

Comment:
Should verify that this should be a `ValueError` and not a `TypeError`.  If `ValueError` is correct, feel free to just resolve this issue.